src/Word.jl

```diff
@@ -0,0 +1,68 @@
+abstract type Word{T} <: AbstractVector{T} end
+
+
+
+#interface for morphism T->Int
+function to_int{T}(t::T) end
+#interface for morphism Int->T
+function from_int{T}(t::T) end
+#interface for ordering on T
+function <=(a::T,b::T) end
+#empty string as monoid neutral element
+Base.one(::Type{Word{T}}) where {T}  = Word(Vector{T}())
+#overload for <= that implements word ordering
+function <=(a::Word{T},b::Word{T})
+#overload for * as concatenation
+function *(a::Word{T},b::Word{T}) where {T} Word(a.ptrs+b.ptrs) end
+#interface for function that returns actual vector inside word
+function data(a::Word{T})
+
+Base.length(w::Word{T}) where {T}  = length(w.ptrs)
```

Comment:
`length` is fine, but for vectors/matrices/nd-arrays it's better to implement `Base.size(w::Word) = size(w.ptrs)`

Reply:
in general have a look at https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array-1
it may seem a little bit cryptic, at first but all it says is that for a subtype of `AbstractArray` we need to implement
`size`, `getindex`, `setindex` (both with the appropriate number of args), and `IndexStyle` → `IndexLinear()` for it to behave like a vector.

most of the standard stuff (like e.g. iteration) will work automatically